### ISSUE TITLE
[ORC] Memory manager in C-API

### ISSUE THREAD

**Issue by wjakob** [llvm/llvm-project#59222]
Dear LLVM team,

I am currently in the looking into transitioning a C-API MCJIT application to ORCv2.

With MCJIT, it was possible to use `LLVMCreateSimpleMCJITMemoryManager()` to set up callbacks to precisely control the placement of executable and data sections during linking.

My application is using this to create relocatable code that can be written to disk (a JIT cache) so that it can be subsequently loaded, `mmap()`-ed into memory and finally run without using LLVM.

I haven't been able to figure out how to do this with ORCv2, in particular when using the provided C-API bindings. I would be grateful for any advice!

Many thanks in advance!

**Comment by wjakob** [comment]
I've been reading up on the different ORCv2 examples and watched some video presentations by @lhames, but have not been able to make progress. Any advice would be greatly appreciated.

**Comment by wjakob** [comment]
Thank you for the quick response @lhames!

> When running ORC on top of RuntimeDyld (MCJIT's memory manager) you could reuse LLVMCreateSimpleMCJITMemoryManager, and we'd just need to add a new RTDyldObjectLinkingLayer construction function to take your custom allocator as an argument.

This sounds like the easier path to follow of the two you mentioned.

My project dynamically loads LLVM via `dlopen` it needs to JIT something, and it uses the C-API to be able to dynamically talk to many different versions of LLVM. In any case, LLVM is a quite big project, and simply running the JIT once comes with certain overheads. When a computation is in any case already cached on disk, it is much faster to just `mmap()` a linked relocatable binary into memory without involving LLVM.

**Comment by lhames** [comment]
> This sounds like the easier path to follow of the two you mentioned.
 
It's definitely the easy option for now. Eventually we will require a move to JITLink (it much more flexible, and required to support new features), but RuntimeDyld will be kept around for some time to enable transition.

> LLVM is a quite big project, and simply running the JIT once comes with certain overheads. When a computation is in any case already cached on disk, it is much faster to just mmap() a linked relocatable binary into memory without involving LLVM.

Certainly compiling from IR is expensive. I'm currently optimizing for a middle ground: (1) A library refactor so that JIT clients only have to depend on the JIT linkers, rather than all of LLVM, and (2) A way to close the JIT linkers without evicting JITd code up-front. The idea is that you would cache relocatable objects and pay the cost of re-linking each time, but not recompiling. This allows you to mix and match lazily-compiled and cached code (with the potential memory layout changes that that implies).

In your case you could still write a custom allocator to try to cache the fully-linked code -- the new APIs won't help with that, but hopefully won't get in the way too much either.

**Comment by wjakob** [comment]
With a looming MCJIT deprecation at some point in the future, my hope is to find a solution with long-term stability.

> Eventually we will require a move to JITLink (it much more flexible, and required to support new features), but RuntimeDyld will be kept around for some time to enable transition.

With that extra context, I am changing my mind and lean towards the latter option you mentioned (A C API for the JITLink memory manager and an ObjectLinkingLayer construction function) even if it is more complicated.

**Comment by lhames** [comment]
Hi @wjakob. I don't think that that's safe: `RTDyldObjectLinkingLayer` will call that callback for each object, and after the first I think that you will see a double-free.

I'm working on a general solution for `RTDyldObjectLinkingLayer` now. I should be able to post a review for you to test shortly.

**Comment by wjakob** [comment]
Got it. Thank you in advance 🙇  -- I am happy to test anything you send my way.

**Comment by lhames** [comment]
@wjakob Posting as-is, since I have to head out for the night: https://reviews.llvm.org/D139393

I don't have a C-based memory manager handy to test it with -- if you don't mind doing some initial work porting yours and working through the bugs with me it would be a big help.

**Comment by wjakob** [comment]
Awesome, thanks so much @lhames! I got my application to the point where it is able to compile and run one JIT-ed kernel before running into other problems (see further questions below). But first, here are two changes that I had to make to your diff:  one field wasn't copied, and it seems to me that requiring non-NULL `CreateContextCtx` is not needed. In my application I don't use this payload.

```cpp
diff --git a/llvm/lib/ExecutionEngine/Orc/OrcV2CBindings.cpp b/llvm/lib/ExecutionEngine/Orc/OrcV2CBindings.cpp
index 62333a49c73f..dcb7dc3cdf32 100644
--- a/llvm/lib/ExecutionEngine/Orc/OrcV2CBindings.cpp
+++ b/llvm/lib/ExecutionEngine/Orc/OrcV2CBindings.cpp
@@ -1095,6 +1095,7 @@ LLVMOrcCreateRTDyldObjectLinkingLayerWithMCJITMemoryManagerLikeCallbacks(
 
     MCJITMemoryManagerLikeCallbacks(MCJITMemoryManagerLikeCallbacks &&Other) {
       std::swap(CreateContextCtx, Other.CreateContextCtx);
+      std::swap(CreateContext, Other.CreateContext);
       std::swap(NotifyTerminating, Other.NotifyTerminating);
       std::swap(AllocateCodeSection, Other.AllocateCodeSection);
       std::swap(AllocateDataSection, Other.AllocateDataSection);
@@ -1159,7 +1160,6 @@ LLVMOrcCreateRTDyldObjectLinkingLayerWithMCJITMemoryManagerLikeCallbacks(
   };
 
   assert(ES && "ES must not be null");
-  assert(CreateContextCtx && "CreateContextCtx must not be null");
   assert(CreateContext && "CreateContext must not be null");
   assert(NotifyTerminating && "NotifyTerminating must not be null");
   assert(AllocateCodeSection && "AllocateCodeSection must not be null");diff --git a/llvm/lib/ExecutionEngine/Orc/OrcV2CBindings.cpp b/llvm/lib/ExecutionEngine/Orc/OrcV2CBindings.cpp
index 62333a49c73f..dcb7dc3cdf32 100644
--- a/llvm/lib/ExecutionEngine/Orc/OrcV2CBindings.cpp
+++ b/llvm/lib/ExecutionEngine/Orc/OrcV2CBindings.cpp
@@ -1095,6 +1095,7 @@ LLVMOrcCreateRTDyldObjectLinkingLayerWithMCJITMemoryManagerLikeCallbacks(
 
     MCJITMemoryManagerLikeCallbacks(MCJITMemoryManagerLikeCallbacks &&Other) {
       std::swap(CreateContextCtx, Other.CreateContextCtx);
+      std::swap(CreateContext, Other.CreateContext);
       std::swap(NotifyTerminating, Other.NotifyTerminating);
       std::swap(AllocateCodeSection, Other.AllocateCodeSection);
       std::swap(AllocateDataSection, Other.AllocateDataSection);
@@ -1159,7 +1160,6 @@ LLVMOrcCreateRTDyldObjectLinkingLayerWithMCJITMemoryManagerLikeCallbacks(
   };
 
   assert(ES && "ES must not be null");
-  assert(CreateContextCtx && "CreateContextCtx must not be null");
   assert(CreateContext && "CreateContext must not be null");
   assert(NotifyTerminating && "NotifyTerminating must not be null");
   assert(AllocateCodeSection && "AllocateCodeSection must not be null");
```

I have two questions:

1. It wasn't clear to me what the `CreateContext` callback was supposed to do. I simply return NULL, and that seems to work. Is that just making the payload for the other callbacks?

2. In my application, I use LLJIT to compile one kernel, at which point everything in there should be cleared out so that it is as good as new. Completely rebuilding the LLJIT is probably overkill, but now I am not sure what interface to use.

There is both `LLVMOrcJITDylibClear` and `LLVMOrcLLJITAddLLVMIRModuleWithRT` that seem to serve related purposes. Do I need one or both, or yet something else?

Many thanks in advance!

**Comment by wjakob** [comment]
Dear @lhames,

did you have a chance to look at my example? It's a modification of another hello-world style example that JITs a function that computes an addition. The only difference is that the application takes care of the memory mapping (i.e. suitable calls to `mmap` and `mprotect`). Besides the POSIX code path, I also added code that should work on Windows.

I'm happy to make changes if you see any issues.

Basically my goal is to get this binding merged soon so that my project can switch to ORCv2 with the next LLVM version.

Many thanks,
Wenzel

**Comment by lhames** [comment]
Not yet -- my apologies.

I aim to take a look tomorrow.

**Comment by lhames** [comment]
This looks perfect! Thank you very much @wjakob!

With your permission I'll commit it with some minor changes (e.g. I'll exercise the Ctx and  CtxCtx arguments, just to make sure that they behave as expected) along with the C API updates.

What name and email would you like me to use for attribution?

**Comment by wjakob** [comment]
Hi @lhames - great, I am happy to hear it. You have my permission to use this code in any way (including merging it into LLVM). For attribution, you can use “Wenzel Jakob <wenzel.jakob@epfl.ch>”

**Comment by lhames** [comment]
Should be fixed in 10d0bd44ae15. Thanks again for your help with this @wjakob!

**Comment by wjakob** [comment]
Amazing, an early Christmas present! Many thanks for landing this 🎉
